go/viperutil/internal/sync/sync.go

```diff
@@ -48,16 +49,28 @@ type Viper struct {
 	subscribers    []chan<- struct{}
 	watchingConfig bool
 
+	fs afero.Fs
+
 	setCh chan struct{}
+
+	// for testing purposes only
+	onConfigWrite func()
+}
+
+func (v *Viper) SetFs(fs afero.Fs) {
+	v.fs = fs
+	v.disk.SetFs(fs)
 }
 
 // New returns a new synced Viper.
 func New() *Viper {
 	return &Viper{
-		disk:  viper.New(),
-		live:  viper.New(),
-		keys:  map[string]*sync.RWMutex{},
-		setCh: make(chan struct{}, 1),
+		disk:          viper.New(),
+		live:          viper.New(),
+		keys:          map[string]*sync.RWMutex{},
+		fs:            afero.NewOsFs(), // default Fs used by viper, but we need this set so loadFromDisk doesn't accidentally nil-out the live fs
+		setCh:         make(chan struct{}, 1),
+		onConfigWrite: func() {},
```

Comment:
Any reason not to set it to nil and only execute it if it's not nil? We should probably have a nil check anyway as a user could set the value to nil, no?

Reply:
It's not possible for a user to set it (it's a private field and this is an internal package), but I was surprised to learn just _how much_ faster it is to do the nil check than an empty func call:

<details><summary>trivial benchmark</summary>

```go
func NewT1(f func()) *T1 {
	if f == nil {
		f = func() {}
	}

	return &T1{f: f}
}

type T1 struct {
	f func()
}

func (t1 *T1) Do() {
	t1.f()
}

type T2 struct {
	f func()
}

func NewT2(f func()) *T2 {
	return &T2{f: f}
}

func (t2 *T2) Do() {
	if t2.f != nil {
		t2.f()
	}
}

func BenchmarkEmptyFunc(b *testing.B) {
	t1 := NewT1(nil)
	for i := 0; i < b.N; i++ {
		t1.Do()
	}
}

func BenchmarkNilCheck(b *testing.B) {
	t2 := NewT2(nil)
	for i := 0; i < b.N; i++ {
		t2.Do()
	}
}

// BenchmarkEmptyFunc-10    	571605042	         2.077 ns/op	       0 B/op	       0 allocs/op
// BenchmarkNilCheck-10     	1000000000	         0.3177 ns/op	       0 B/op	       0 allocs/op
```

</details>

soooo lemme make that change!